File: Difficulty: Medium/Bitwise AND of the Array/bitwise-and-of-the-array.cpp
//{ Driver Code Starts
//Initial Template for C++
#include<bits/stdc++.h>
using namespace std;

// } Driver Code Ends

// User function Template for C++

class Solution {
    public:
    int count(int N, vector<int> A, int X) {
        // Finding the maximum element in the array
        int maxElement = -1;
        for (int i = 0; i < N; i++) {
            maxElement = max(maxElement, A[i]);
        }

        // Determine the number of bits required to represent maxElement and X
        int maxBits = log2(maxElement);
        int bitsInX = log2(X);

        // Initialize two possible answers to N (maximum possible count)
        int minExcluded1 = N, minExcluded2 = N;

        // Checking bits from (bitsInX + 1) to maxBits
        for (int i = bitsInX + 1; i <= maxBits; i++) {
            int excludeCount = 0;

            // Counting numbers that do not have the current bit set
            for (int j = 0; j < N; j++) {
                if ((A[j] & (1 << i)) == 0) {
                    excludeCount++;
                }
            }
            
            // Updating minimum excluded count
            minExcluded1 = min(minExcluded1, excludeCount);
        }

        // To track which elements are already excluded
        vector<int> visited(N, 0);
        int excludedCount = 0;

        // Iterating over bits of X from highest to lowest
        for (int i = bitsInX; i >= 0; i--) {
            if (X & (1 << i)) {
                // If the bit is set in X, exclude elements where this bit is not set
                for (int j = 0; j < N; j++) {
                    if ((A[j] & (1 << i)) == 0 && visited[j] == 0) {
                        visited[j] = 1;
                        excludedCount++;
                    }
                }
            } else {
                // If the bit is not set in X, find an alternative exclusion count
                int tempExcludeCount = 0;
                for (int j = 0; j < N; j++) {
                    if ((A[j] & (1 << i)) == 0 && visited[j] == 0) {
                        tempExcludeCount++;
                    }
                }
                minExcluded2 = min(minExcluded2, excludedCount + tempExcludeCount);
            }
        }

        // Return the minimum of both exclusion approaches
        return min(minExcluded2, minExcluded1);
    }
};


//{ Driver Code Starts.
int main() {
    int t;
    cin>>t;
    while(t--)
    {
        int N,X;
        cin>>N>>X;
        vector<int> A(N);
        for(auto &i:A)
            cin>>i;
        Solution obj;
        int ans = obj.count(N, A, X);
        cout<<ans<<endl;
    
cout << "~" << "\n";
}
}
// } Driver Code Ends

File: Difficulty: Medium/Bitwise AND of the Array/bitwise-and-of-the-array.java
//{ Driver Code Starts
//Initial Template for Java
import java.io.*;
import java.util.*; 
class GFG{
    public static void main(String args[]) throws IOException { 
        BufferedReader read = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(read.readLine());
        
        while(t-- > 0){
            String input_line[] = read.readLine().trim().split("\\s+");
            int N = Integer.parseInt(input_line[0]);
            int X = Integer.parseInt(input_line[1]);
            input_line = read.readLine().trim().split("\\s+");
            int A[]= new int[N];
            for(int i = 0; i < N; i++)
                A[i] = Integer.parseInt(input_line[i]);
            Solution ob = new Solution();
            int ans = ob.count(N, A, X); 
            System.out.println(ans);
        
System.out.println("~");
}
    } 
}
// } Driver Code Ends

// User function Template for Java

class Solution { 
    int count(int N, int A[], int X) {   
        int prefix = 0, minRemovals = N;

        // Iterate over each bit position from most significant (30) to least significant (0)
        for (int bit = 30; bit >= 0; bit--) {
            // If the current bit in X is set (1), update prefix and continue
            if (((X >> bit) & 1) != 0) {
                prefix ^= (1L << bit);
                continue;
            }

            int countValid = 0;
            long potentialPrefix = prefix ^ (1L << bit);

            // Count how many elements in A satisfy the condition
            for (int j = 0; j < N; j++) {
                if ((A[j] & potentialPrefix) == potentialPrefix) {
                    countValid++;
                }
            }

            // Update the minimum number of elements to remove
            minRemovals = Math.min(minRemovals, N - countValid);
        }

        return minRemovals;
    }
}  
